pkg/collector/clock.go

```diff
@@ -0,0 +1,156 @@
+// Copyright 2024 VMware, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package collector
+
+import (
+	"sync"
+	"time"
+)
+
+// timer allows for injecting fake or real timers into code that needs to do arbitrary things based
+// on time. We do not include the C() method, as we only support timers created with AfterFunc.
+type timer interface {
+	Stop() bool
+	Reset(d time.Duration) bool
+}
+
+// clock allows for injecting fake or real clocks into code that needs to do arbitrary things based
+// on time. We only support a very limited interface at the moment, with only the methods required
+// by CollectingProcess.
+type clock interface {
+	Now() time.Time
+	AfterFunc(d time.Duration, f func()) timer
+}
+
+// realClock implements the clock interface using functions from the time package.
+type realClock struct{}
+
+func (realClock) Now() time.Time {
+	return time.Now()
+}
+
+func (realClock) AfterFunc(d time.Duration, f func()) timer {
+	return time.AfterFunc(d, f)
+}
+
+type fakeTimer struct {
+	targetTime time.Time
+	f          func()
+	clock      *fakeClock
```

Comment:
The clock here is for deleting itself from the parent clock in convenience?

Reply:
It's not a convenience, it's required for functionality (implementation of `Stop`). But yes, it is so the timer can remove itself from the clock's list of timers.